### ISSUE TITLE
ValueError: TRIGGER has invalid parameters

### ISSUE THREAD

**Issue by aceisace** [C4ptainCrunch/ics.py#160]
@C4ptainCrunch 
Hi, I've been trying to use ics.py for parsing events from an ical URL (in my case yahoo) for one of my projects ([project link](https://github.com/aceisace/E-Paper-Calendar-with-iCal-sync-and-live-weather)). I always get the same error on days where I have events, but I can't figure out why. Could you please help with debugging this issue?

Here is a copy of the error:
```
Traceback (most recent call last):
  File "/home/pi/E-Paper-Master/Calendar/stable.py", line 38, in <module>
    c = Calendar(urlopen(url).read().decode())
  File "/usr/local/lib/python3.5/dist-packages/ics/icalendar.py", line 69, in __init__
    self._populate(container[0])  # Use first calendar
  File "/usr/local/lib/python3.5/dist-packages/ics/component.py", line 60, in _populate
    extractor.function(self, lines)  # Send a list or empty list
  File "/usr/local/lib/python3.5/dist-packages/ics/icalendar.py", line 214, in events
    calendar.events = list(map(event_factory, lines))
  File "/usr/local/lib/python3.5/dist-packages/ics/icalendar.py", line 213, in event_factory
    return Event._from_container(x, tz=calendar._timezones)
  File "/usr/local/lib/python3.5/dist-packages/ics/component.py", line 32, in _from_container
    k._populate(container)
  File "/usr/local/lib/python3.5/dist-packages/ics/component.py", line 60, in _populate
    extractor.function(self, lines)  # Send a list or empty list
  File "/usr/local/lib/python3.5/dist-packages/ics/event.py", line 497, in alarms
    event.alarms = list(map(alarm_factory, lines))
  File "/usr/local/lib/python3.5/dist-packages/ics/event.py", line 496, in alarm_factory
    return af._from_container(x)
  File "/usr/local/lib/python3.5/dist-packages/ics/component.py", line 32, in _from_container
    k._populate(container)
  File "/usr/local/lib/python3.5/dist-packages/ics/component.py", line 63, in _populate
    extractor.function(self, lines[0])  # Send the element
  File "/usr/local/lib/python3.5/dist-packages/ics/alarm.py", line 209, in trigger
    raise ValueError('TRIGGER has invalid parameters')
ValueError: TRIGGER has invalid parameters
```
Thanks in advance.

**Comment by C4ptainCrunch** [comment]
Hi !
Could you attach a calendar that is making the lib fail ?
The best would be a minimal example so i'm able to reproduce the error while not being drowned in noise :)

**Comment by aceisace** [comment]
@C4ptainCrunch 
I've saved my (yahoo) iCalendar as a text file. Is there any way I can directly send the file to you? I'm just asking because iCalendars contain some personal information as well. If that's not possible, I can attach a section of the iCalendar showing an event which is causing the error.

**Comment by C4ptainCrunch** [comment]
Here is a minimal example derived from the file @aceisace sent me by email
```
BEGIN:VCALENDAR
PRODID:redacted
VERSION:2.0
BEGIN:VEVENT
SUMMARY:redacted
DTSTART:20190101T000000
DTEND:20190101T000000
BEGIN:VALARM
ACTION:DISPLAY
TRIGGER;RELATED=START:-PT1H
END:VALARM
END:VEVENT
END:VCALENDAR
```

**Comment by aceisace** [comment]
@C4ptainCrunch 
Thanks for taking a look at the problem-causing iCalendar file. I'm sorry for sending that file too late. It seems it got stuck in my outgoing folder first.

**Comment by dkudlek** [comment]
@C4ptainCrunch

Hi, 
I also have an problem with VALARM and the TRIGGER field. That's why I add it to this issue. I get a parsing error for an event with a positive TRIGGER with a missing sign (which was not exported from my google calendar). 

# Error Message:
```
Traceback (most recent call last):
File "calendar_backend.py", line 35, in <module>
ical = Calendar(decode)
File "C:\Tools\Anaconda3\lib\site-packages\ics\icalendar.py", line 68, in __init__
self._populate(container[0])  # Use first calendar
File "C:\Tools\Anaconda3\lib\site-packages\ics\component.py", line 59, in _populate
extractor.function(self, lines)  # Send a list or empty list
File "C:\Tools\Anaconda3\lib\site-packages\ics\icalendar.py", line 207, in events
calendar.events = set(map(event_factory, lines))
File "C:\Tools\Anaconda3\lib\site-packages\ics\icalendar.py", line 206, in event_factory
return Event._from_container(x, tz=calendar._timezones)
File "C:\Tools\Anaconda3\lib\site-packages\ics\component.py", line 31, in _from_container
k._populate(container)
File "C:\Tools\Anaconda3\lib\site-packages\ics\component.py", line 59, in _populate
extractor.function(self, lines)  # Send a list or empty list
File "C:\Tools\Anaconda3\lib\site-packages\ics\event.py", line 556, in alarms
event.alarms = list(map(alarm_factory, lines))
File "C:\Tools\Anaconda3\lib\site-packages\ics\event.py", line 555, in alarm_factory
return af._from_container(x)
File "C:\Tools\Anaconda3\lib\site-packages\ics\component.py", line 31, in _from_container
k._populate(container)
File "C:\Tools\Anaconda3\lib\site-packages\ics\component.py", line 62, in _populate
extractor.function(self, lines[0])  # Send the element
File "C:\Tools\Anaconda3\lib\site-packages\ics\alarm.py", line 189, in trigger
alarm.trigger = parse_duration(line.value[1:])
File "C:\Tools\Anaconda3\lib\site-packages\ics\utils.py", line 126, in parse_duration
raise parse.ParseError()
ics.parse.ParseError
```

# Event: 
The event from my google calender which causes the problem.

```
BEGIN:VEVENT
DTSTART;VALUE=DATE:20180806
DTEND;VALUE=DATE:20180807
DTSTAMP:20190816T202741Z
UID:6dhj4c3474rj2b9p6lhm2b9k6ormcb9pccsjab9m61gj2dpn6kr3eor564@google.com
CREATED:20180729T055121Z
DESCRIPTION:Filiale hat bis 17:30 geöffnet
LAST-MODIFIED:20180729T055232Z
LOCATION:Postbank-Finanzcenter
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Paket abholen \, 17:00
TRANSP:OPAQUE
BEGIN:VALARM
ACTION:DISPLAY
DESCRIPTION:This is an event reminder
TRIGGER:P0DT9H0M0S
END:VALARM
BEGIN:VALARM
ACTION:DISPLAY
DESCRIPTION:This is an event reminder
TRIGGER:-P0DT4H0M0S
END:VALARM
BEGIN:VALARM
ACTION:DISPLAY
DESCRIPTION:This is an event reminder
TRIGGER:-P1DT15H0M0S
END:VALARM
END:VEVENT
```


# Problem:
After debugging I noticed that ``TRIGGER:P0DT9H0M0S`` from the event above is the issue. It is an allday event which has a alarm on the same day at 09:00 (9 hours after the event started). It seems that google does not export the ``+`` sign for a reminder after the start of an event. This causes an issue in ``alarm.py:189`` where the expected sign is cut off. Which in turn causes ``parse_duration`` to fail as it receives ``0DT9H0M0S`` missing the ``P``.  

```python
@Alarm._extracts('TRIGGER', required=True)
def trigger(alarm, line):
    if not line.params or 'DURATION' in line.params.get('VALUE', ''):
        alarm.trigger = parse_duration(line.value[1:]) 
```

I fixed it by adding a guard for the sign. Is this in line with the intended behavior though?

```python
@Alarm._extracts('TRIGGER', required=True)
def trigger(alarm, line):
    if not line.params or 'DURATION' in line.params.get('VALUE', ''):
      if line.value[0] in '+-':
          alarm.trigger = parse_duration(line.value[1:])
      else:
          alarm.trigger = parse_duration(line.value[0:]) 
```

**Comment by C4ptainCrunch** [comment]
@dkudlek The problem here is that we should not skip the first character with `line.value[1:]`.
`parse_duration()` is capable of dealing with the leading `+` or `-` (or having neither).
This will be fixed int #183.

**Comment by dkudlek** [comment]
@C4ptainCrunch This is what I also tried at first. If I remember correctly, it was throwing another error then. ... Something something... duration can not be negative?

Skipping the leading sign means taking the positive absolute value. That's why I made my fix in the trigger function.

But I may have also updated to a newer version of ics inbetween testing.

I'll look forward to the fix. 

Thx.

**Comment by C4ptainCrunch** [comment]
That error about the negative trigger was also incorrect :/
It will be fixed in #183 too :)
